README.txt
===========================================================================================
CONTENTS
===========================================================================================
1.  About
2.  Supported formats
3.  Supported languages
4.  Features
5.  Archive contents
6.  How to build library
7.  Warning
8.  DISCLAIMER
9.  COPYRIGHT
10. MP3 and AAC LICENSE
11. Author


===========================================================================================
1. ABOUT
===========================================================================================
libZPlay ver. 2.02 (Win32)

Version: 2.02
Date: May, 2010. 

This is multimedia library for playing mp3, mp2, mp1, ogg, flac, oga, ac3, aac, wav and pcm files and streams. 

This library integrate all decoders for supported formats. You don't need additional libraries.
Library is using native WINAPI functions from kernel32.dll, user32.dll, gdi32.dll and winmm.dll.
All these libraries are standard libraries on all Microsoft Windows. There is no need for MFC or .NET support.
Just pure WINAPI. 

Library will play music directly to soundcard.
Simple and easy. No need for additional programming.
You can play music using only 3 lines of code (create class, open file and start playing).
Simple, simple, simple ... 


libZPlay library is combination of multiple decoding/encoding libraries.
Read topic 10. COPYRIGHT for more informations.

 
===========================================================================================
2. Supported formats
=========================================================================================== 
 
- MP3  -> mp3, mp2, mp1  MPEG-1, MPEG-2, MPEG 2.5 - Layer I, Layer II, Layer III.
          Read topic 9: MP3 and AAC LICENSE 
 
- OGG  -> ogg  VORBIS OGG  

- AC3  -> A/52 aka. ac3

- AAC  -> HE, LC, MAIN and LTP profile, MPEG2 and MPEG-4 AAC decoder.
          Read topic 9: MP3 and AAC LICENSE

- FLAC -> flac, oga  Native FLAC and OGG flac 
 
- WAVE -> wav  PCM WAVE with RIFF header, uncompressed, 8, 16 bit per sample  

- PCM  -> any file  Raw PCM audio format, uncomressed, 8 16 bit per sample  

 
===========================================================================================
3. Supported languages
===========================================================================================

Binary distribution is customized for these languages:

- Standard C/C++ (native) 
- Visual C++ .NET (VC.NET wrapper class) 
- Visual Basic .NET (VB.NET wrapper class) 
- C# .NET (C#.NET wrapper class) 
- Delphi (Delphi wrapper class) 
- and any other language which can use exported __stdcall functions from dll library 
 

libZPlay library is written in C++ and C++ interface is native interface for this library.
But there are wrapper classes for Visual C++ .NET, Visual Basic .NET, C# .NET and Delphi.  
 
===========================================================================================
4. Features
===========================================================================================

- supports mp3, ogg, flac, ac3, aac, wav, pcm 
- supports wave-in input (get samples from line-in, microphone, ....)
- supports encoding samples into mp3, ogg, flac, wave and aac streams
- UNICODE support for functions using strings ( filename, error messages, ID3 info, ... ) 
- read info tags from file ( ID3v1, ID3v2, VORBIS comment, wave RIFF list, FLAC VORBIS comment, embeded picture ) 
- easy drawing embeded picture to screen
- REVERSE PLAYING (you can play song backward in real time, "real" backward playing without predecoding in PCM) 
- built in equaliter, user can specify own subbands, add subbands, ... 
- built in echo and reverb sound processor 
- pitch/rate/tempo change 
- stereo cut effect 
- detect beat-per-minute using 2 methods ( peak detection and autocorrelation) 
- loop playing (you can play just one portion of file, repeat this portion ... ) 
- channel mixer (you can mix left and right chanel, convert stereo to mono ... ) 
- volume slide control (fade-in and fade-out) 
- internal volume control (without affecting wave out device volume) 
- pulling VU meter data (you can simply create VU meter) 
- pulling FFT data (you can simply build FFT spectrum analyzer) 
- built in FFT spectrum graph, library will draw complete spectrum graph on your window 
- callback mechanism (class will inform you about events like play, stop, fade, ...) 
- access to decoded PCM data throug callback message 
- play data from disk files, UNICODE SUPPORT 
- play data from allocated memory 
- play dynamic streams (push small memory chunks into stream while song is playing, for download streams) 
- selecting wave output (for multiple waveout devices) 
- easy play, pause, seek ... 
- class interface for Visual C++ .NET, Visual Basic .NET, C# .NET and Delphi. 
 

===========================================================================================
5. ARCHIVE CONTENTS
===========================================================================================

.\
  decoders\
      a52dec     - liba52 project and source files
      faac       - FAAC library (AAC encoder)
      faad       - FAAD library (AAC decoder)
      jpeglib    - The Independent JPEG Group's JPEG software (IJG) library
      lame       - LAME library (mp3 encoder)
      libFLAC    - libFLAC library (FLAC encoder/decoder)
      libmad     - libmad library (mp3 decoder)
      libogg     - libogg library
      libvorbis  - libvorbis library (vorbis encoder/decoder)
      lpng       - libpng library (png picture decoder)
      soundtouch - SoundTouch library
      zlib       - zlib library

  doc_src        - documentation source code, need Doc-O-Matic to build documentation
  example        - C++ examples
  image          - images

  include        - libZPlay include file for C++
  license        - license informations for encoder/decoder libraries
  src            - libZPlay source files
  vc++           - libZPlay VC++ project and solution files

===========================================================================================
6. HOW TO BUILD LIBRARY
===========================================================================================
Library is written in C++.
To build this library you need MSVC++ 6.0 + SP6 + Processor Pack or later.
Binary distribution is compiled with Visual C++ 2005 Express Edition.

There are 2 types of libZPlay library:
- full version
- PF (patent free) version, without mp3 and aac support

Open .\vc++\libzplay.dsw or .\vc++\libzplay.sln and build libzplay_dynamic or libzplay_pf_dynamic.


===========================================================================================
7. WARNING 
===========================================================================================
Documentation for VB.NET, Delphi and C# was generated from C++ documentation.
Function declarations are tested and there is all OK. But comments,
examples and parametes explanations can contain C++ syntax code or some other strange
expressions because I didn't read all this.
I just adapted function declarations.  
 

===========================================================================================
8. DISCLAIMER
=========================================================================================== 
This library is distributed as freeware.
This software is provided "as is", without any guarantee made as to its suitability or
fitness for any particular use.
It may contain bugs, so use of this tool is at your own risk.
Author takes no responsibility for any damage that may unintentionally be caused through its use.  
 
===========================================================================================
9. COPYRIGHT 
===========================================================================================
libZPlay is released under GNU GPL license. 

NOTE: Read topic 11: MP3 and AAC license.

 *
 * Supported by:
 *
 * ============================================================================
 * libmad - MPEG audio decoder library
 * Copyright (C) 2000-2004 Underbit Technologies, Inc. <support@underbit.com>
 * http://www.underbit.com/products/mad/
 * GNU General Public License ( GPL.TXT )
 * ============================================================================
 * THE OggVorbis SOURCE CODE IS (C) COPYRIGHT 1994-2002 
 * by the Xiph.Org Foundation http://www.xiph.org/
 * BSD-STYLE SOURCE LICENSE  ( XIPH.TXT )
 * ============================================================================
 * LIBA52
 * free library for decoding ATSC A/52 streams.
 * It is released under the terms of the GPL license.
 * ============================================================================
 * FLAC - Free Lossless Audio Codec
 * Copyright (C) 2001,2002,2003,2004,2005,2006,2007  Josh Coalson
 * http://flac.sourceforge.net/
 *
 * This file is part the FLAC project.  FLAC is comprised of several
 * components distributed under difference licenses.  The codec libraries
 * are distributed under Xiph.Org's BSD-like license (see the file
 * XIPH.TXT in this distribution).  All other programs, libraries, and
 * plugins are distributed under the LGPL or GPL (see LGPL.TXT and
 * GPL.TXT).  The documentation is distributed under the Gnu FDL (see
 * FDL.TXT).  Each file in the FLAC distribution contains at the top the
 * terms under which it may be distributed.
 *
 * Since this particular file is relevant to all components of FLAC,
 * it may be distributed under the Xiph.Org license, which is the least
 * restrictive of those mentioned above.  See the file XIPH.TXT in this
 * distribution.
 * ============================================================================
 * FAAD2 - Freeware Advanced Audio (AAC) Decoder including SBR decoding
 * Copyright (C) 2003-2005 M. Bakker, Nero AG, http://www.nero.com
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 * Any non-GPL usage of this software or parts of this software is strictly
 * forbidden.
 *
 * The "appropriate copyright message" mentioned in section 2c of the GPLv2
 * must read: "Code from FAAD2 is copyright (c) Nero AG, www.nero.com"
 *
 * Commercial non-GPL licensing of this software is possible.
 * For more info contact Nero AG through Mpeg4AAClicense@nero.com.
 *
 * ============================================================================
 * FAAC - Freeware Advanced Audio Coder Copyright (C) 2001 M. Bakker
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version. 
 *
 * This library is distributed in the hope that it will be useful,

ChangeLog.txt

===========================================================================================
LIBZPLAY HISTORY
===========================================================================================


  Version 2.02                     May 2010.
  -------------------------------------------------------------
  - fixed problem with ID3v1 algorithm
  - fixed problem with callback function (blocking thread)



  Version 2.00                     April 2010.
  -------------------------------------------------------------
  - fixed memory leaks
  - fixed problem with ID3v2 algorithm
  - added support for extended ID3 fields(composer, encoder, ...)
  - added support for embeded pictures(jpeg, png, bmp) in tags(ID3v2, VORBIS COMMENT, FLAC VORBIS COMMENT)
  - added support for sfWaveIn format (read samples from line-in, microphine, ...)
    and play these samples to wave-out with specifed effects.
  - added support for encoding (mp3, ogg, flac, aac)
 

  Version 1.14                     April 2010.
  -------------------------------------------------------------
  - fixed crash bug in ID3v2 tag algorithm
  - fixed memory leak when open function fails on mp3 file
  - fixed error in .NET interface (wrong layout of ID3 fields) 


  Version 1.12                     April 2010.
  -------------------------------------------------------------
  - added MP3 and AAC license informations

  Version 1.11                     April 2010.
  -------------------------------------------------------------
  - PF (patent free) version of version 1.12
    This is version 1.12 without MP3 and AAC support



  Version 1.04                     April 2010.
  -------------------------------------------------------------
  - fixed reverse playing on mp3 stream. Sorry for this bug.
    This bug wasn't present in version 1.01 but I made some
    testing and debugging, and unwanted code wasn't deleted.
    And there is, bug in version 1.02 and 1.03. Sorry for this.
    Now is fixed. 
    

  Version 1.03                    April 2010.
  -------------------------------------------------------------
  + added AAC ADTS support
  - fixed crash bug on Windows Vista and Windows 7 for .NET interface


  Version 1.02                    March 2010.
  -------------------------------------------------------------
  - fix mp3 dynamic stream management (loss of samples)
  - fix support for FLAC OGG (added spearate sfFLACOgg format)
  + added A/52 aka AC-3 support
  + added Spectrum FFT graph type
  + added FFT graph color management and background bitmap


  Version 1.01                    March 2010.
  -------------------------------------------------------------
  - first release


src/wfft.h
/*
 *  libzplay - windows ( WIN32 ) multimedia library for playing mp3, mp2, ogg, wav, flac and raw PCM streams
 *  FFT class
 *
 *  Copyright (C) 2003-2009 Zoran Cindori ( zcindori@inet.hr )
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *
 * ver: 1.14
 * date: 15. April, 2010.
 *
 *	SUPPORTED BY:
 * ============================================================================
 * FFT part is a routine made by Mr.Ooura. This routine is a freeware.
 * Contact Mr.Ooura for details of distributing licenses.
 * http://momonga.t.u-tokyo.ac.jp/~ooura/fft.html
 *
*/

#ifndef _W_FFT_H_
#define _W_FFT_H_


#ifndef HIGHPREC
#define REAL_IS_FLOAT
#else
#define REAL_IS_DOUBLE
#endif

#if 0
#undef REAL_IS_FLOAT
#undef REAL_IS_DOUBLE
#undef REAL_IS_LONGDOUBLE
#endif

/*
Fast Fourier/Cosine/Sine Transform
    dimension   :one
    data length :power of 2
    decimation  :frequency
    radix       :split-radix
    data        :inplace
    table       :use
functions
    cdft: Complex Discrete Fourier Transform
    rdft: Real Discrete Fourier Transform
    ddct: Discrete Cosine Transform
    ddst: Discrete Sine Transform
    dfct: Cosine Transform of RDFT (Real Symmetric DFT)
    dfst: Sine Transform of RDFT (Real Anti-symmetric DFT)
function prototypes
    void cdft(int, int, REAL *, int *, REAL *);
    void rdft(int, int, REAL *, int *, REAL *);
    void ddct(int, int, REAL *, int *, REAL *);
    void ddst(int, int, REAL *, int *, REAL *);
    void dfct(int, REAL *, REAL *, int *, REAL *);
    void dfst(int, REAL *, REAL *, int *, REAL *);


-------- Complex DFT (Discrete Fourier Transform) --------
    [definition]
        <case1>
            X[k] = sum_j=0^n-1 x[j]*exp(2*pi*i*j*k/n), 0<=k<n
        <case2>
            X[k] = sum_j=0^n-1 x[j]*exp(-2*pi*i*j*k/n), 0<=k<n
        (notes: sum_j=0^n-1 is a summation from j=0 to n-1)
    [usage]
        <case1>
            ip[0] = 0; // first time only
            cdft(2*n, 1, a, ip, w);
        <case2>
            ip[0] = 0; // first time only
            cdft(2*n, -1, a, ip, w);
    [parameters]
        2*n            :data length (int)
                        n >= 1, n = power of 2
        a[0...2*n-1]   :input/output data (REAL *)
                        input data
                            a[2*j] = Re(x[j]), 
                            a[2*j+1] = Im(x[j]), 0<=j<n
                        output data
                            a[2*k] = Re(X[k]), 
                            a[2*k+1] = Im(X[k]), 0<=k<n
        ip[0...*]      :work area for bit reversal (int *)
                        length of ip >= 2+sqrt(n)
                        strictly, 
                        length of ip >= 
                            2+(1<<(int)(log(n+0.5)/log(2))/2).
                        ip[0],ip[1] are pointers of the cos/sin table.
        w[0...n/2-1]   :cos/sin table (REAL *)
                        w[],ip[] are initialized if ip[0] == 0.
    [remark]
        Inverse of 
            cdft(2*n, -1, a, ip, w);
        is 
            cdft(2*n, 1, a, ip, w);
            for (j = 0; j <= 2 * n - 1; j++) {
                a[j] *= 1.0 / n;
            }
        .


-------- Real DFT / Inverse of Real DFT --------
    [definition]
        <case1> RDFT
            R[k] = sum_j=0^n-1 a[j]*cos(2*pi*j*k/n), 0<=k<=n/2
            I[k] = sum_j=0^n-1 a[j]*sin(2*pi*j*k/n), 0<k<n/2
        <case2> IRDFT (excluding scale)
            a[k] = (R[0] + R[n/2]*cos(pi*k))/2 + 
                   sum_j=1^n/2-1 R[j]*cos(2*pi*j*k/n) + 
                   sum_j=1^n/2-1 I[j]*sin(2*pi*j*k/n), 0<=k<n
    [usage]
        <case1>
            ip[0] = 0; // first time only
            rdft(n, 1, a, ip, w);
        <case2>
            ip[0] = 0; // first time only
            rdft(n, -1, a, ip, w);
    [parameters]
        n              :data length (int)
                        n >= 2, n = power of 2
        a[0...n-1]     :input/output data (REAL *)
                        <case1>
                            output data
                                a[2*k] = R[k], 0<=k<n/2
                                a[2*k+1] = I[k], 0<k<n/2
                                a[1] = R[n/2]
                        <case2>
                            input data
                                a[2*j] = R[j], 0<=j<n/2
                                a[2*j+1] = I[j], 0<j<n/2
                                a[1] = R[n/2]
        ip[0...*]      :work area for bit reversal (int *)
                        length of ip >= 2+sqrt(n/2)
                        strictly, 
                        length of ip >= 
                            2+(1<<(int)(log(n/2+0.5)/log(2))/2).
                        ip[0],ip[1] are pointers of the cos/sin table.
        w[0...n/2-1]   :cos/sin table (REAL *)
                        w[],ip[] are initialized if ip[0] == 0.
    [remark]
        Inverse of 
            rdft(n, 1, a, ip, w);
        is 
            rdft(n, -1, a, ip, w);
            for (j = 0; j <= n - 1; j++) {
                a[j] *= 2.0 / n;
            }
        .


-------- DCT (Discrete Cosine Transform) / Inverse of DCT --------
    [definition]
        <case1> IDCT (excluding scale)
            C[k] = sum_j=0^n-1 a[j]*cos(pi*j*(k+1/2)/n), 0<=k<n
        <case2> DCT
            C[k] = sum_j=0^n-1 a[j]*cos(pi*(j+1/2)*k/n), 0<=k<n
    [usage]
        <case1>
            ip[0] = 0; // first time only
            ddct(n, 1, a, ip, w);
        <case2>
            ip[0] = 0; // first time only
            ddct(n, -1, a, ip, w);
    [parameters]
        n              :data length (int)
                        n >= 2, n = power of 2
        a[0...n-1]     :input/output data (REAL *)
                        output data
                            a[k] = C[k], 0<=k<n
        ip[0...*]      :work area for bit reversal (int *)
                        length of ip >= 2+sqrt(n/2)
                        strictly, 
                        length of ip >= 
                            2+(1<<(int)(log(n/2+0.5)/log(2))/2).
                        ip[0],ip[1] are pointers of the cos/sin table.
        w[0...n*5/4-1] :cos/sin table (REAL *)
                        w[],ip[] are initialized if ip[0] == 0.
    [remark]
        Inverse of 
            ddct(n, -1, a, ip, w);
        is 
            a[0] *= 0.5;
            ddct(n, 1, a, ip, w);
            for (j = 0; j <= n - 1; j++) {
                a[j] *= 2.0 / n;
            }
        .


-------- DST (Discrete Sine Transform) / Inverse of DST --------
    [definition]
        <case1> IDST (excluding scale)
            S[k] = sum_j=1^n A[j]*sin(pi*j*(k+1/2)/n), 0<=k<n
        <case2> DST
            S[k] = sum_j=0^n-1 a[j]*sin(pi*(j+1/2)*k/n), 0<k<=n
    [usage]
        <case1>
            ip[0] = 0; // first time only
            ddst(n, 1, a, ip, w);
        <case2>
            ip[0] = 0; // first time only
            ddst(n, -1, a, ip, w);
    [parameters]
        n              :data length (int)
                        n >= 2, n = power of 2
        a[0...n-1]     :input/output data (REAL *)
                        <case1>
                            input data
                                a[j] = A[j], 0<j<n
                                a[0] = A[n]
                            output data
                                a[k] = S[k], 0<=k<n
                        <case2>
                            output data
                                a[k] = S[k], 0<k<n
                                a[0] = S[n]
        ip[0...*]      :work area for bit reversal (int *)
                        length of ip >= 2+sqrt(n/2)
                        strictly, 
                        length of ip >= 
                            2+(1<<(int)(log(n/2+0.5)/log(2))/2).
                        ip[0],ip[1] are pointers of the cos/sin table.
        w[0...n*5/4-1] :cos/sin table (REAL *)
                        w[],ip[] are initialized if ip[0] == 0.
    [remark]
        Inverse of 
            ddst(n, -1, a, ip, w);
        is 
            a[0] *= 0.5;
            ddst(n, 1, a, ip, w);
            for (j = 0; j <= n - 1; j++) {
                a[j] *= 2.0 / n;
            }
        .


-------- Cosine Transform of RDFT (Real Symmetric DFT) --------
    [definition]
        C[k] = sum_j=0^n a[j]*cos(pi*j*k/n), 0<=k<=n
    [usage]
        ip[0] = 0; // first time only
        dfct(n, a, t, ip, w);
    [parameters]
        n              :data length - 1 (int)
                        n >= 2, n = power of 2
        a[0...n]       :input/output data (REAL *)
                        output data
                            a[k] = C[k], 0<=k<=n
        t[0...n/2]     :work area (REAL *)
        ip[0...*]      :work area for bit reversal (int *)
                        length of ip >= 2+sqrt(n/4)
                        strictly, 
                        length of ip >= 
                            2+(1<<(int)(log(n/4+0.5)/log(2))/2).
                        ip[0],ip[1] are pointers of the cos/sin table.
        w[0...n*5/8-1] :cos/sin table (REAL *)
                        w[],ip[] are initialized if ip[0] == 0.
    [remark]
        Inverse of 
            a[0] *= 0.5;
            a[n] *= 0.5;
            dfct(n, a, t, ip, w);
        is 
            a[0] *= 0.5;
            a[n] *= 0.5;
            dfct(n, a, t, ip, w);
            for (j = 0; j <= n; j++) {
                a[j] *= 2.0 / n;
            }
        .


-------- Sine Transform of RDFT (Real Anti-symmetric DFT) --------
    [definition]
        S[k] = sum_j=1^n-1 a[j]*sin(pi*j*k/n), 0<k<n
    [usage]
        ip[0] = 0; // first time only
        dfst(n, a, t, ip, w);
    [parameters]
        n              :data length + 1 (int)
                        n >= 2, n = power of 2
        a[0...n-1]     :input/output data (REAL *)
                        output data
                            a[k] = S[k], 0<k<n
                        (a[0] is used for work area)
        t[0...n/2-1]   :work area (REAL *)
        ip[0...*]      :work area for bit reversal (int *)
                        length of ip >= 2+sqrt(n/4)
                        strictly, 
                        length of ip >= 
                            2+(1<<(int)(log(n/4+0.5)/log(2))/2).
                        ip[0],ip[1] are pointers of the cos/sin table.
        w[0...n*5/8-1] :cos/sin table (REAL *)
                        w[],ip[] are initialized if ip[0] == 0.
    [remark]
        Inverse of 
            dfst(n, a, t, ip, w);
        is 
            dfst(n, a, t, ip, w);
            for (j = 1; j <= n - 1; j++) {
                a[j] *= 2.0 / n;
            }
        .


Appendix :
    The cos/sin table is recalculated when the larger table required.
    w[] and ip[] are compatible with all routines.
*/




#include <stdlib.h>
#include <math.h>

#ifdef REAL_IS_FLOAT
	typedef float REAL;

	#define SIN(x) sinf(x)
	#define COS(x) cosf(x)

	#define CDFT_RECURSIVE_N 1024
#endif

#ifdef REAL_IS_DOUBLE
	typedef double REAL;

	#define SIN(x) sin(x)
	#define COS(x) cos(x)

	#define CDFT_RECURSIVE_N 512
#endif

#ifdef REAL_IS_LONGDOUBLE
	typedef long double REAL;

	#define SIN(x) sinl(x)
	#define COS(x) cosl(x)

	#define CDFT_RECURSIVE_N 512
#endif

#define PI 3.1415926535897932384626433832795029L
#define PI2 6.283185307179586476925286766559L



void cdft(int n, int isgn, REAL *a, int *ip, REAL *w);
void rdft(int n, int isgn, REAL *a, int *ip, REAL *w);
void ddct(int n, int isgn, REAL *a, int *ip, REAL *w);
void ddst(int n, int isgn, REAL *a, int *ip, REAL *w);
void dfct(int n, REAL *a, REAL *t, int *ip, REAL *w);
void dfst(int n, REAL *a, REAL *t, int *ip, REAL *w);



void fft(REAL *data, long size);
void ifft(REAL* data, long size);
void realfft_split(REAL *data,long n);
void irealfft_split(REAL *data,long n);


// FFT WINDOW
#define  FFT_WINDOW_RECTANGULAR 1
#define  FFT_WINDOW_HAMMING 2
#define  FFT_WINDOW_HANN 3	
#define  FFT_WINDOW_COSINE	4
#define  FFT_WINDOW_LANCZOS 5	
#define  FFT_WINDOW_BARTLETT 6
#define  FFT_WINDOW_TRIANGULAR	7
#define  FFT_WINDOW_GAUSS 8
#define  FFT_WINDOW_BARTLETT_HANN 9
#define  FFT_WINDOW_BLACKMAN 10
#define  FFT_WINDOW_NUTTALL 11
#define  FFT_WINDOW_BLACKMAN_HARRIS 12
#define  FFT_WINDOW_BLACKMAN_NUTTALL 13
#define  FFT_WINDOW_FLAT_TOP 14

class WFFT {
public:
	WFFT();
	~WFFT();

// =========================================================================================================================
	// initialize class
	int Initialize(unsigned int nFFTPoints, unsigned int nSampleRate, unsigned int nChannel, unsigned int nBitPerSample);

	//	PARAMETERS:
	//		nFFTPoints
	//			Number of FFT points. Must be power of 2.
	//
	//		nSampleRate
	//			Sample rate.
	//
	//		nChannel
	//			Number of channels. Must be 1 or 2
	//
	//		nBitPerSample
	//			Bit per sample. Must be 8, 16 or 32.
	//
	//	RETURN VALUES:
	//		1	- all OK
	//		0	- error
// ==========================================================================================================================
	// get number of harmonics
	unsigned int GetNumberOfHarmonics();

	//	PARAMETER:
	//		None.
	//
	//	RETURN VALUES:
	//		Number of harmonics.
// ==========================================================================================================================
	// get harmonics frequencies
	int GetHarmonicFreqReal(REAL *pHarmonicFreq);

	//	PARAMETERS:
	//		pHarmonicFreq
	//			Pointer to buffer receiving harmonic frequencies. User must allocate this buffer.
	//			Call GetNumberOfHarmonics to determine number of harmonics and then allocate memory for this buffer.
	//			If this parameter is NULL, function returns number of harmonics.
	//
	//	RETURN VALUES:
	//		Function returns number of harmonics.
	//		If pHarmonicFreq is NULL, function returns number of harmonics.
// ==========================================================================================================================
	// get harmonics frequencies
	int GetHarmonicFreqInt(int *pHarmonicFreq);

	//	PARAMETERS:
	//		pHarmonicFreq
	//			Pointer to buffer receiving harmonic frequencies. User must allocate this buffer.
	//			Call GetNumberOfHarmonics to determine number of harmonics and then allocate memory for this buffer.
	//			If this parameter is NULL, function returns number of harmonics.
	//
	//	RETURN VALUES:
	//		Number of harmonics.
	//		If pHarmonicFreq is NULL, function returns number of harmonics.
// ==========================================================================================================================

	// set FFT window
	int SetWindow(unsigned int nWindow);
	//
	//	PARAMETERS:
	//		nWindow
	//			FFT window.
	//
	//	RETURN VALUES:
	//		1	- all OK
	//		0	- error
// =======================================

	int GetWindow();

// ==========================================================================================================================
	// free initialized memory
	int Free();

	//	PARAMETERS:
	//		None.
	//
	//	RETURN VALUES:
	//		1	- all OK
	//		0	- error
// ==========================================================================================================================
	// calculate FFT

	int CalculateFFT(void *pSamples, unsigned int nNumberOfSamples);

	// PARAMETERS:
	//		pSamples
	//			Pointer to buffer with PCM samples.
	//
	//		nNumberOfSamples
	//			Number of samples in buffer.
	//
	//	RETURN VALUES:
	//		1	- all OK
	//		0	- error
// =========================================================================================================================	
	// get amplitudes from FFT
	int GetAmplitudeReal(REAL *pLeftAmplitude, REAL *pRightAmplitude);

	//	PARAMETERS:
	//		pLeftAmplitude
	//			Pointer to buffer receiving amplitudes of left channel harmonics.
	//			User must allocate this memory. Use GetNumberOfHarmonics to determine number of
	//			harmonics and allocate enough memory.
	//			If this value is NULL, function will not fill this buffer.
	//
	//		pRightAmplitude
	//			Pointer to buffer receiving amplitudes of right channel harmonics.
	//			User must allocate this memory. Use GetNumberOfHarmonics to determine number of
	//			harmonics and allocate enough memory.
	//			If this value is NULL, function will not fill this buffer.
	//
	//	RETURN VALUES:
	//		1	- all OK
	//		0	- error

// =========================================================================================================================	
	// get amplitudes from FFT
	int GetAmplitudeInt(int *pLeftAmplitude, int *pRightAmplitude);

	//	PARAMETERS:
	//		pLeftAmplitude
	//			Pointer to buffer receiving amplitudes of left channel harmonics.
	//			User must allocate this memory. Use GetNumberOfHarmonics to determine number of
	//			harmonics and allocate enough memory.
	//			If this value is NULL, function will not fill this buffer.
	//
	//		pRightAmplitude
	//			Pointer to buffer receiving amplitudes of right channel harmonics.
	//			User must allocate this memory. Use GetNumberOfHarmonics to determine number of
	//			harmonics and allocate enough memory.
	//			If this value is NULL, function will not fill this buffer.
	//
	//	RETURN VALUES:
	//		1	- all OK
	//		0	- error
// ==========================================================================================================================
	// get FFT phase
	int GetPhaseReal(REAL *pLeftPhase, REAL *pRightPhase);

	//	PARAMETERS:
	//		pLeftPhase
	//			Pointer to buffer receiving phase of left channel harmonics.
	//			User must allocate this memory. Use GetNumberOfHarmonics to determine number of
	//			harmonics and allocate enough memory.
	//			If this value is NULL, function will not fill this buffer.
	//		
	//		pRightPhase
	//			Pointer to buffer receiving phase of left channel harmonics.
	//			User must allocate this memory. Use GetNumberOfHarmonics to determine number of
	//			harmonics and allocate enough memory.
	//			If this value is NULL, function will not fill this buffer.
	//
	//	RETURN VALUES:
	//		1	- all OK
	//		0	- error	

// ==========================================================================================================================
	// get FFT phase
	int GetPhaseInt(int *pLeftPhase, int *pRightPhase);

	//	PARAMETERS:
	//		pLeftPhase
	//			Pointer to buffer receiving phase of left channel harmonics.
	//			User must allocate this memory. Use GetNumberOfHarmonics to determine number of
	//			harmonics and allocate enough memory.
	//			If this value is NULL, function will not fill this buffer.
	//		
	//		pRightPhase
	//			Pointer to buffer receiving phase of left channel harmonics.
	//			User must allocate this memory. Use GetNumberOfHarmonics to determine number of
	//			harmonics and allocate enough memory.
	//			If this value is NULL, function will not fill this buffer.
	//
	//	RETURN VALUES:
	//		1	- all OK
	//		0	- error	

// ==========================================================================================================================


	void *GetSamplesBuffer();



// ==========================================================================================================================
private:
	unsigned int c_nFFTPoints;		// number of fft points
	REAL c_SqrtFFTPoints;			// sqrt of FFT points
	unsigned int c_nHarmonicNumber;	// number of harmonics
	int c_nFFTWindow;

	unsigned int c_nSampleRate;		// samplerate
	unsigned int c_nChannel;		// number of channels ( 1, 2 )
	unsigned int c_nBitPerSample;	// bits per sample ( 8, 16, 32 )

	REAL *c_window;					// window
	REAL *c_harmonics;				// frequency of each harmonic
	REAL *c_samplesl;				// samples
	REAL *c_samplesr;				// samples
	REAL *c_amplitude;				// amplitude of harmonics
	REAL *c_phase;					// phase of harmonic
	int *c_bit_reversal;			// working area for bit reversal
	REAL *c_cos_sin_table;			// cos/sin table


	int _alloc();
	int _free();

	static void _convertFFTtoAmplitudeReal(unsigned int FFTPoints, unsigned int SqrtFFTPoints, REAL *fft, REAL *amplitude);
	static void _convertFFTtoAmplitudeInt(unsigned int FFTPoints, unsigned int SqrtFFTPoints, REAL *fft, int *amplitude);

	static void _convertFFTtoPhaseReal(unsigned int FFTPoints, unsigned int SqrtFFTPoints, REAL *fft, REAL *phase);
	static void _convertFFTtoPhaseInt(unsigned int FFTPoints, unsigned int SqrtFFTPoints, REAL *fft, int *phase);

};


#endif
